GAE/src/org/waterforpeople/mapping/app/web/RawDataRestServlet.java

```diff
@@ -143,35 +144,64 @@ protected RestResponse handleRequest(RestRequest req) throws Exception {
-                    answer.setSurveyInstanceId(instance.getKey().getId());
-                    answer.setSurveyId(s.getKey().getId());
-                    answer.setCollectionDate(instance.getCollectionDate());
+            for (Entry<Long, Map<Integer, String[]>> responseEntry : incomingResponses
+                    .entrySet()) {
+                Long questionId = responseEntry.getKey();
+                Map<Integer, String[]> iterationMap = responseEntry.getValue();
+
+                for (Entry<Integer, String[]> iterationEntry : iterationMap.entrySet()) {
+                    Integer iteration = iterationEntry.getKey();
+                    String response = iterationEntry.getValue()[0];
+                    String type = iterationEntry.getValue()[1];
+                    QuestionAnswerStore answer = null;
+                    if (existingAnswers.containsKey(questionId)) {
+                        if (existingAnswers.get(questionId).containsKey(iteration)) {
+                            answer = existingAnswers.get(questionId).get(iteration);
+                        }
+                    }
+
+                    // New answer/iteration
+                    if (answer == null) {
+                        answer = new QuestionAnswerStore();
+                        answer.setQuestionID(questionId.toString());
+                        answer.setSurveyInstanceId(instance.getKey().getId());
+                        answer.setSurveyId(s.getKey().getId());
+                        answer.setCollectionDate(instance.getCollectionDate());
+                    }
+
+                    answer.setValue(response);
+                    answer.setType(type);
+                    updatedAnswers.add(answer);
                 }
-                answer.setValue(qasEntry.getValue()[0]);
-                answer.setType(qasEntry.getValue()[1]);
-                updatedAnswers.add(answer);
             }
+
             qasDao.save(updatedAnswers);
 
             // remove entities with no updated response
             List<QuestionAnswerStore> deletedAnswers = new ArrayList<QuestionAnswerStore>();
-            for (Long qasId : existingAnswers.keySet()) {
-                if (!incomingAnswers.containsKey(qasId)) {
-                    deletedAnswers.add(existingAnswers.get(qasId));
+
+            for (Long questionId : existingAnswers.keySet()) {
+                for (Integer iteration : existingAnswers.get(questionId).keySet()) {
+                    if (incomingResponses.containsKey(questionId)) {
+                        if (!incomingResponses.get(questionId).containsKey(iteration)) {
+                            // Iteration has been deleted
+                            deletedAnswers.add(existingAnswers.get(questionId).get(iteration));
+                        }
+                    } else {
+                        // All iterations has been deleted
+                        deletedAnswers.addAll(existingAnswers.get(questionId).values());
```

Comment:
I was thinking we should maybe remove the `else` branch as it may simply be the case that the column with that particular question was deleted from the excel sheet, which doesn't necessarily mean the intention was to delete the data from the backend.  Perhaps we can just ignore if we don't find the question id